fuse_optimizers/src/variable_stamp_index.cpp

```diff
@@ -47,19 +47,29 @@ namespace fuse_optimizers
 {
 ros::Time VariableStampIndex::currentStamp() const
 {
-  auto compare_stamps = [](const StampedMap::value_type& lhs, const StampedMap::value_type& rhs)
-  {
-    return lhs.second < rhs.second;
-  };
-  auto iter = std::max_element(stamped_index_.begin(), stamped_index_.end(), compare_stamps);
-  if (iter != stamped_index_.end())
-  {
-    return iter->second;
+  if(!unique_stamps_.empty()){
+    return *unique_stamps_.end();
   }
-  else
-  {
-    return ros::Time(0, 0);
+  return ros::Time(0, 0);
+}
+
+ros::Time VariableStampIndex::firstStamp() const
+{
+  if(!unique_stamps_.empty()){
+    return *unique_stamps_.begin();
   }
+  return ros::Time(0, 0);
+} 
+
+ros::Time VariableStampIndex::ithStamp(size_t idx) const{
```

Comment:
What are your thoughts on the proper action if `idx` exceeds the size of the container? Stdlib avoids the issue and just claims it is undefined behavior. We could do that as well and blindly call `std::advance()` without checking sizes first. Or we could throw an exception to provide some debugging information. Or we can return Time(0) as some sort of magic code for "does not exist".

Reply:
The previous behaviour for firstStamp would return Time(0) for the failure cases so I would stick with this for the rest. However I see how Time(0) makes sense for firstStamp and even currentStamp when the time index is empty. I think time(0) should be returned when empty, and an exception should be thrown if index is greater than the size.